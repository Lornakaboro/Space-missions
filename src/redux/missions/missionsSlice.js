import { createSlice } from '@reduxjs/toolkit';
import getMissions from './missionsThunk';

const initialState = {
  missions: [],
  loading: false,
  error: null,
};

const missionHandlerReducer = (state, { payload }) => ({
  ...state,
  missions: state.missions.map((mission) => (mission.mission_id === payload
    ? { ...mission, reserved: !mission.reserved }
    : mission)),
});

const missionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    missionHandler: missionHandlerReducer,
  },
  extraReducers: (builder) => {
    builder
      .addCase(getMissions.pending, (state) => ({
        ...state,
        loading: true,
      }))
      .addCase(getMissions.fulfilled, (state, { payload }) => ({
        ...state,
        loading: false,
        missions: payload,
        error: null,
      }))
      .addCase(getMissions.rejected, (state, { payload }) => ({
        ...state,
        error: payload,
        loading: false,
      }));
  },
});

export const { missionHandler } = missionSlice.actions;

export default missionSlice.reducer;
