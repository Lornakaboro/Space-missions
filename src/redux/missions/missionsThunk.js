/* eslint-disable  camelcase */

import { createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const url = 'https://api.spacexdata.com/v3/missions';

const formatMissions = (missions) => missions.map(({ mission_name, mission_id, description }) => ({
  mission_name,
  mission_id,
  description,
}));

export default createAsyncThunk(
  'missions/getMissions',
  async (name, { rejectWithValue, fulfillWithValue }) => {
    try {
      const { data } = await axios.get(url);
      const missionData = formatMissions(data);
      return fulfillWithValue(missionData);
    } catch (error) {
      return rejectWithValue({ ...error.response.data });
    }
  },
);
