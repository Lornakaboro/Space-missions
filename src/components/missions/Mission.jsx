import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { missionHandler } from '../../redux/missions/missionsSlice';

const Mission = ({ mission }) => {
  const {
    id, name, description, reserved,
  } = mission;
  const dispatch = useDispatch();

  return (
    <tr className="mission-item">
      <td className="mission-name">{name}</td>
      <td className="mission-decription">{description}</td>
      {reserved ? (
        <>
          <td className="join-status">
            <span className="reserved">Active Member</span>
          </td>
          <td className="mission-action">
            <button
              className="leave-mission"
              type="button"
              onClick={() => dispatch(missionHandler(id))}
            >
              Leave Mission
            </button>
          </td>
        </>
      ) : (
        <>
          <td className="join-status">
            <span className="unreserved">NOT A MEMBER</span>
          </td>
          <td className="mission-action">
            <button
              className="join-mission"
              type="button"
              onClick={() => dispatch(missionHandler(id))}
            >
              Join Mission
            </button>
          </td>
        </>
      )}
    </tr>
  );
};

Mission.propTypes = {
  mission: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reserved: PropTypes.bool,
  }).isRequired,
};

export default Mission;
