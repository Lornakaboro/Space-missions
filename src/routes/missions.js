import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Mission from '../components/missions/Mission';
import getMissions from '../redux/missions/missionsThunk';

const Missions = () => {
  const dispatch = useDispatch();

  const { missions } = useSelector((store) => store.missions);

  useEffect(() => {
    dispatch(getMissions());
  }, [dispatch]);

  return (
    <section>
      <table>
        <thead>
          <tr>
            <th className="t-name">Mission</th>
            <th className="t-description">Description</th>
            <th className="t-status">Status</th>
            <th className="t-action"> </th>
          </tr>
        </thead>
        <tbody>
          {missions.map((item) => (
            <Mission key={item.mission_id} mission={item} />
          ))}
        </tbody>
      </table>
    </section>
  );
};

export default Missions;
