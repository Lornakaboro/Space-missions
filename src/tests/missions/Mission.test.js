import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { useDispatch } from 'react-redux';
import Mission from '../../src/components/missions/Mission';
import { missionHandler } from '../../src/redux/missions/missionsSlice';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('axios', () => ({
  get: jest.fn(() =>
    Promise.resolve({
      data: [
        {
          mission_id: 'a1',
          mission_name: 'Test Mission',
          description: 'This is a test mission',
          reserved: false,
        },
      ],
    }),
  ),
}));

describe('Mission', () => {
  const mission = {
    mission_id: 'a1',
    mission_name: 'Test Mission',
    description: 'This is a test mission',
    reserved: false,
  };

  it('Mission component renders correctly', () => {
    const tree = renderer
      .create(
        <table>
          <thead>
            <tr>
              <th className="t-name">Mission</th>
              <th className="t-description">Description</th>
              <th className="t-status">Status</th>
              <th className="t-action"> </th>
            </tr>
          </thead>
          <tbody>
            <Mission mission={mission} />
          </tbody>
        </table>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders mission details correctly', () => {
    render(
      <table>
        <thead>
          <tr>
            <th className="t-name">Mission</th>
            <th className="t-description">Description</th>
            <th className="t-status">Status</th>
            <th className="t-action"> </th>
          </tr>
        </thead>
        <tbody>
          <Mission mission={mission} />
        </tbody>
      </table>,
    );
    const missionName = screen.getByText('Test Mission');
    const missionDescription = screen.getByText('This is a test mission');
    expect(missionName).toBeInTheDocument();
    expect(missionDescription).toBeInTheDocument();
  });

  it('dispatches missionHandler action on join/leave button click', () => {
    const mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);

    const { getByRole } = render(
      <table>
        <thead>
          <tr>
            <th className="t-name">Mission</th>
            <th className="t-description">Description</th>
            <th className="t-status">Status</th>
            <th className="t-action"> </th>
          </tr>
        </thead>
        <tbody>
          <Mission mission={mission} />
        </tbody>
      </table>,
    );
    const joinButton = getByRole('button', { name: 'Join Mission' });

    fireEvent.click(joinButton);
    expect(mockDispatch).toHaveBeenCalledWith(missionHandler('a1'));

    fireEvent.click(joinButton);
    expect(mockDispatch).toHaveBeenCalledWith(missionHandler('a1'));
  });
});
