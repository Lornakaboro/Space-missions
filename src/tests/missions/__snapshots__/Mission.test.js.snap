// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission Mission component renders correctly 1`] = `
<table>
  <thead>
    <tr>
      <th
        className="t-name"
      >
        Mission
      </th>
      <th
        className="t-description"
      >
        Description
      </th>
      <th
        className="t-status"
      >
        Status
      </th>
      <th
        className="t-action"
      >
         
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="mission-item"
    >
      <td
        className="mission-name"
      >
        Test Mission
      </td>
      <td
        className="mission-decription"
      >
        This is a test mission
      </td>
      <td
        className="join-status"
      >
        <span
          className="unreserved"
        >
          NOT A MEMBER
        </span>
      </td>
      <td
        className="mission-action"
      >
        <button
          className="join-mission"
          onClick={[Function]}
          type="button"
        >
          Join Mission
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
